ably/rest_client.go

```diff
@@ -140,24 +144,88 @@ func (c *RestClient) do(r *Request) (*http.Response, error) {
 	}
 	resp, err := c.opts.httpclient().Do(req)
 	if err != nil {
-		return nil, newError(50000, err)
+		return nil, newError(ErrInternalError, err)
 	}
 	resp, err = c.handleResponse(resp, r.Out)
-	switch {
-	case err == nil:
-		return resp, nil
-	case code(err) == 40140:
-		if r.NoRenew || !c.Auth.isTokenRenewable() {
-			return nil, err
-		}
-		if _, err := c.Auth.reauthorise(); err != nil {
-			return nil, err
+	if err != nil {
+		if e, ok := err.(*Error); ok {
+			if canFallBack(e.StatusCode) &&
+				(c.opts.FallbackHostsUseDefault ||
+					strings.HasSuffix(req.URL.Host, defaultOptions.RestHost)) {
+				fallback := defaultOptions.FallbackHosts
+				if c.opts.FallbackHosts != nil {
+					fallback = c.opts.FallbackHosts
+				}
+				if len(fallback) > 0 {
+					// we copy the hosts to avoid accidentally mutating the array while
+					// running.
+					left := make([]string, len(fallback))
+					copy(left, fallback)
```

Comment:
The `copy` is unnecessary since slices are value types: simply doing `left := fallback` will create a copy of the `fallback` slice header value. When you assign `left = n` below, it'll overwrite the slice header in `left` but leave `fallback` intact. This is OK because you're not mutating the _contents_ of the slice, only the header

Reply:
`do()` is getting a bit long, so you might want to break out the host picking code into a function/method of its own

Reply:
@ORBAT  I don't think there will be benefit of breaking down `do` I prefer it this way because even if we move host picking code into a function it is only going to be used here so inlining the logic makes sense and also helps when reading and comparing notes to the spec. I addressed the other comment.